sampo/scheduler/timeline/zone_timeline.py

```diff
@@ -0,0 +1,238 @@
+from collections import deque
+
+from sortedcontainers import SortedList
+
+from sampo.schemas.requirements import ZoneReq
+from sampo.schemas.time import Time
+from sampo.schemas.types import EventType, ScheduleEvent
+from sampo.schemas.zones import ZoneConfiguration, Zone, ZoneTransition
+from sampo.utilities.collections_util import build_index
+
+
+class ZoneTimeline:
+
+    def __init__(self, config: ZoneConfiguration):
+        def event_cmp(event: ScheduleEvent | Time | tuple[Time, int, int]) -> tuple[Time, int, int]:
+            if isinstance(event, ScheduleEvent):
+                if event.event_type is EventType.INITIAL:
+                    return Time(-1), -1, event.event_type.priority
+
+                return event.time, event.seq_id, event.event_type.priority
+
+            if isinstance(event, Time):
+                # instances of Time must be greater than almost all ScheduleEvents with same time point
+                return event, Time.inf().value, 2
+
+            if isinstance(event, tuple):
+                return event
+
+            raise ValueError(f'Incorrect type of value: {type(event)}')
+
+        self._timeline = {zone: SortedList([ScheduleEvent(-1, EventType.INITIAL, Time(0), None, status)],
+                                           key=event_cmp)
+                          for zone, status in config.start_statuses.items()}
+        self._config = config
+
+    def find_min_start_time(self, zones: list[ZoneReq], parent_time: Time, exec_time: Time):
+        # here we look for the earliest time slot that can satisfy all the zones
+
+        start = parent_time
+        scheduled_wreqs: list[ZoneReq] = []
+
+        type2status: dict[str, int] = build_index(zones, lambda w: w.kind, lambda w: w.required_status)
+
+        queue = deque(zones)
+
+        i = 0
+        while len(queue) > 0:
+            # if i > 0 and i % 50 == 0:
+            #     print(f'Warning! Probably cycle in looking for time slot for all reqs: {i} iteration')
+            #     print(f'Current queue size: {len(queue)}')
```

Comment:
rudiment comments

Reply:
These lines should be uncommented when performance issues rise